### PR TITLE
Add graphical SYCL demos

### PR DESCRIPTION
**Pull request by Vtec234**

### CODE REVIEW

**Comment by DuncanMcBain** [comment]
Also, the README should be modified to mention the demo folder and what it has in it

File: demos/game_of_life/sim.hpp [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+      auto width = this->m_width;
+      auto height = this->m_height;
+
+      cgh.parallel_for<class gameoflifesimkernel>(
+          // Work on each cell in parallel
+          cl::sycl::range<2>(width, height), [=](cl::sycl::item<2> item) {
+            size_t x = item.get(0);
+            size_t y = item.get(1);
+
+            // Unrolled loop for accessing the 3x3 stencil
```
Comment (Ruyk) [inline demos/game_of_life/sim.hpp]: do you really need to unroll the loop? have you profiled this and check there is any benefit?

File: demos/game_of_life/sim.hpp [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+                       8.0f;
+            auto new_vel = (rv[item] + vel) / 2.0f;
+            wv[item] = new_vel;
+
+            // Increase values to have brighter colours
+            new_vel = sycl::fabs(new_vel) * 5.0f + sycl::float2(0.2f, 0.2f);
+
+            // Set image pixel to new colour decided by state and "velocity"
+            img[sycl::id<2>(y, x)] = sycl::cl_uchar4(
+                float(int(new_state)) * new_vel.x() * 255.0f, 0,
```
Comment (Ruyk) [inline demos/game_of_life/sim.hpp]: use explicit casts

File: demos/include/codeplay_demo.hpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+// Adds logo drawing functionality.
+class CodeplayDemoApp : public ci::app::App {
+ protected:
+  ci::gl::Texture2dRef m_codeplay_tex =
+      ci::gl::Texture2d::create(loadImage(loadAsset("logo.png")));
+
+  void draw_codeplay_logo() {
+    float tex_ratio =
+        float(m_codeplay_tex->getWidth()) / float(m_codeplay_tex->getHeight());
```
Comment (Ruyk) [inline demos/include/codeplay_demo.hpp]: use explicit casts

File: demos/include/double_buf.hpp [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+template <typename T>
+class DoubleBuf {
+ public:
+  template <typename... U>
+  DoubleBuf(U&&... vals)
+      : m_read_a_and_not_b(true),
+        m_a(std::forward<U>(vals)...),
+        m_b(std::forward<U>(vals)...) {}
+
+  void swap() { m_read_a_and_not_b = !m_read_a_and_not_b; }
```
Comment (Ruyk) [inline demos/include/double_buf.hpp]: m_read_a should be enough.

is it not better to have an enum class with USE_A or USE_B members?

File: demos/include/double_buf.hpp [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+// Double-buffers any kind of value.
+template <typename T>
+class DoubleBuf {
+ public:
+  template <typename... U>
+  DoubleBuf(U&&... vals)
+      : m_read_a_and_not_b(true),
+        m_a(std::forward<U>(vals)...),
+        m_b(std::forward<U>(vals)...) {}
```
Comment (Ruyk) [inline demos/include/double_buf.hpp]: I guess m_a and m_b can only be buffers

File: demos/include/integrator.hpp [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+ * `step`,
+ * and N+1 values `vals` with the initial conditions of y(N-1), .., y1, y, t, in
+ * that order,
+ * returns the new values of y(N-1), ..., y1, y, t after the step in a tuple.
+ * Uses Runge-Kutta RK4 integration. */
+template <typename time_t, typename func_t, typename... Args>
+std::tuple<Args...> integrate_step_rk4(func_t func, time_t step, Args... vals) {
+  static_assert(sizeof...(Args) >= 2,
+                "Do you want infinite loops in your compiler? Because this is "
+                "how you get infinite loops in your compiler.");
```
Comment (Ruyk) [inline demos/include/integrator.hpp]: please, use a clearer error message and explain why

File: demos/include/integrator.hpp [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+  // calculate the new values from all ks using the formula y_(i+1) = y_i +
+  // step/6 * (k0 + 2*k1 + 2*k2 + k3)
+  auto final = add_tuples(k0s, mult_tuple(k1s, time_t(2)));
+  final = add_tuples(final, mult_tuple(k2s, time_t(2)));
+  final = add_tuples(final, std::move(k3s));
+  final = mult_tuple(final, step / time_t(6));
+  final = add_tuples(final, init);
+
+  return final;
```
Comment (Ruyk) [inline demos/include/integrator.hpp]: did you write this or is this based on some blog post?

File: demos/include/integrator.hpp [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+ * that order,
+ * returns the new values of y(N-1), ..., y1, y, t after the step in a tuple.
+ * Uses Euler integration. */
+template <typename time_t, typename func_t, typename... Args>
+std::tuple<Args...> integrate_step_euler(func_t func, time_t step, Args... vals
+                                         // TODO kmem ptr for temporary storage
+                                         ) {
+  static_assert(sizeof...(Args) >= 2,
+                "Do you want infinite loops in your compiler? Because this is "
+                "how you get infinite loops in your compiler.");
```
Comment (Ruyk) [inline demos/include/integrator.hpp]: error message

File: demos/include/integrator.hpp [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+/* Given a function `func` expressing a derivative of order N, a time step size
+ * `step`,
+ * and N+1 values `vals` with the initial conditions of y(N-1), .., y1, y, t, in
+ * that order,
+ * returns the new values of y(N-1), ..., y1, y, t after the step in a tuple.
+ * Uses Euler integration. */
+template <typename time_t, typename func_t, typename... Args>
+std::tuple<Args...> integrate_step_euler(func_t func, time_t step, Args... vals
+                                         // TODO kmem ptr for temporary storage
```
Comment (Ruyk) [inline demos/include/integrator.hpp]: replace TODO for a github issue

File: demos/include/tuple_utils.hpp [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+
+template <typename tuple_t, typename... Args>
+void setvar(tuple_t const& tpl, Args&... args) {
+  setvar_helper<0>(tpl, args...);
+}
+
+// Calls function with argument tuple.
+namespace {
+template <typename Func, typename Tuple, size_t... Ids>
+AUTO_FUNC(call_help(Func&& func, Tuple&& args, index_sequence<Ids...>),
```
Comment (Ruyk) [inline demos/include/tuple_utils.hpp]: why a macro? just use the standard C++ syntax

File: demos/mandelbrot/main.cpp [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+
+constexpr size_t WIDTH = 800;
+constexpr size_t HEIGHT = 600;
+
+class MandelbrotApp
+#ifdef CODEPLAY_DRAW_LOGO
+    : public CodeplayDemoApp
+#else
+    : public ci::app::App
+#endif
```
Comment (Ruyk) [inline demos/mandelbrot/main.cpp]: why macro

File: demos/mandelbrot/main.cpp [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+{
+  // Use doubles for more zoom
+  MandelbrotCalculator<double> m_calc;
```
Comment (Ruyk) [inline demos/mandelbrot/main.cpp]: note that doubles will cause a major drop in performance in many accelerators , can you make doubles optional? maybe parametrize the execution?

File: demos/mandelbrot/main.cpp [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+#include <CL/sycl.hpp>
+namespace sycl = cl::sycl;
+
+#include <codeplay_demo.hpp>
+
+#include "mandel.hpp"
+
+
+constexpr size_t WIDTH = 800;
+constexpr size_t HEIGHT = 600;
```
Comment (Ruyk) [inline demos/mandelbrot/main.cpp]: add comments explaining the constants. Also use "u" at the end.

File: demos/mandelbrot/main.cpp [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+    this->m_tex = ci::gl::Texture2d::create(nullptr, GL_RGBA, WIDTH, HEIGHT,
+                                            ci::gl::Texture2d::Format()
+                                                .dataType(GL_UNSIGNED_BYTE)
+                                                .internalFormat(GL_RGBA));
+  }
+
+  void update() override {
+    // Transform coordinates from the ones used here - center point
+    // and range - to the ones used in MandelbrotCalculator - min and max X, Y.
+    double range_x = m_range * double(WIDTH) / double(HEIGHT);
```
Comment (Ruyk) [inline demos/mandelbrot/main.cpp]: use explicit casts

File: demos/mandelbrot/mandel.cpp [hunk 15]
```diff
@@ -0,0 +1,39 @@
+/***************************************************************************
+ *
+ *  Copyright (C) 2017 Codeplay Software Limited
```
Comment (Ruyk) [inline demos/mandelbrot/mandel.cpp]: copyright needs update

File: demos/mandelbrot/mandel.hpp [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+        m_q(sycl::default_selector{},
+            [&](sycl::exception_list el) {
+              for (auto const& e : el) {
+                try {
+                  std::rethrow_exception(e);
+                } catch (std::exception const& e) {
+                  std::cout << "SYCL exception caught:\n:" << e.what()
+                            << std::endl;
+                }
+              }
```
Comment (Ruyk) [inline demos/mandelbrot/mandel.hpp]: see my comment about exception handlers

File: demos/mandelbrot/mandel.hpp [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+
+  // Accelerated SYCL queue and storage for image data
+  sycl::queue m_q;
+  sycl::buffer<sycl::cl_uchar4, 2> m_img;
+
+  // Boundaries on the part of the complex plane which we want to view
+  num_t m_minx = -2;
+  num_t m_maxx = 1;
+  num_t m_miny = -1;
+  num_t m_maxy = 1;
```
Comment (Ruyk) [inline demos/mandelbrot/mandel.hpp]: why not use std::pair or create a type for coordinates?

File: demos/mandelbrot/mandel.hpp [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+ private:
+  void internal_calc() {
+    m_q.submit([&](sycl::handler& cgh) {
+      auto img_acc = m_img.get_access<sycl::access::mode::discard_write>(cgh);
+
+      constexpr size_t MAX_ITERS = 500;
+      // Anything above this number is assumed divergent. To do less
+      // computation, this is the _square_ of the maximum absolute value
+      // of a non-divergent number
+      constexpr num_t DIVERGENCE_LIMIT = num_t(256);
```
Comment (Ruyk) [inline demos/mandelbrot/mandel.hpp]: explicit casts

File: demos/mandelbrot/mandel.hpp [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+      // Calculates how many iterations does it take to diverge? MAX_ITERS if in
+      // Mandelbrot set
+      const auto how_mandel = [](num_t re, num_t im) -> num_t {
```
Comment (Ruyk) [inline demos/mandelbrot/mandel.hpp]: if you use east const, be consistent. Otherwise use the west const

File: demos/mandelbrot/mandel.hpp [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+          z_im = num_t(2) * z_re * z_im + im;
+          z_re = z_re2;
+
+          abs_sq = z_re * z_re + z_im * z_im;
+
+          // Branching here isn't ideal, but it's the simplest
+          if (abs_sq >= DIVERGENCE_LIMIT) {
+            num_t log_zn = sycl::log(abs_sq) / num_t(2);
+            num_t nu =
+                sycl::log(log_zn / sycl::log(num_t(2))) / sycl::log(num_t(2));
```
Comment (Ruyk) [inline demos/mandelbrot/mandel.hpp]: why not pre-calculate sycl::log(2)  outside the kernel?

File: demos/mandelbrot/mandel.hpp [hunk 21]
```diff
@@ -0,0 +1,190 @@
...
+
+      // Dummy variable copies to avoid capturing `this` in kernel lambda
+      size_t width = m_width;
+      size_t height = m_height;
+      num_t minx = m_minx;
+      num_t maxx = m_maxx;
+      num_t miny = m_miny;
+      num_t maxy = m_maxy;
+
+      // Use the MandelbrotCalculator class for unique kernel name typ
```
Comment (Ruyk) [inline demos/mandelbrot/mandel.hpp]: typ -> type

File: demos/nbody/main.cpp [hunk 22]
```diff
@@ -0,0 +1,380 @@
...
+
+using num_t = float;
+constexpr num_t PI = num_t(3.141592653589793238462643383279502884197169399);
+
+class NBodyApp
+#ifdef CODEPLAY_DRAW_LOGO
+    : public CodeplayDemoApp
+#else
+    : public ci::app::App
+#endif
```
Comment (Ruyk) [inline demos/nbody/main.cpp]: why macro

File: demos/nbody/main.cpp [hunk 23]
```diff
@@ -0,0 +1,380 @@
...
+{
+  // -- GUI --
+  // Distribution choice
+  enum {
+    UI_DISTRIB_CYLINDER = 0,
+    UI_DISTRIB_SPHERE = 1,
+  };
```
Comment (Ruyk) [inline demos/nbody/main.cpp]: use enum class please

File: demos/nbody/sim.cpp [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+void GravSim<float>::step() {
+  this->internal_step();
+}
+
+/* Doubles disabled.
+template <>
+void GravSim<double>::step() {
+        this->_step();
+}
+*/
```
Comment (Ruyk) [inline demos/nbody/sim.cpp]: why?

File: demos/nbody/sim.hpp [hunk 25]
```diff
@@ -0,0 +1,333 @@
...
+// Handler for queue exceptions
+const auto except_handler = [](sycl::exception_list el) {
+  for (std::exception_ptr const& e : el) {
+    try {
+      std::rethrow_exception(e);
+    } catch (sycl::exception& e) {
+      std::cout << "EXCEPTION:\n" << e.what() << std::endl;
+    }
+  }
+};
```
Comment (Ruyk) [inline demos/nbody/sim.hpp]: move this to a common header file and create a common functor

File: demos/nbody/sim.hpp [hunk 26]
```diff
@@ -0,0 +1,333 @@
...
+      auto wpos = std::get<1>(writes);
+
+      // Dummy variable copies to avoid capturing `this` in kernel lambda
+      num_t t = m_time;
+      size_t n_bodies = m_n_bodies;
+      integrator_t integrator = m_integrator;
+
+      // Launch different kernel depending on the force choice
+      switch (m_force) {
+        case force_t::GRAVITY: {
```
Comment (Ruyk) [inline demos/nbody/sim.hpp]: use function dispatch instead of a switch

File: demos/nbody/sim.hpp [hunk 27]
```diff
@@ -0,0 +1,333 @@
...
+                if (integrator == integrator_t::EULER) {
+                  std::tie(wvel[id], wpos[id], std::ignore) =
+                      integrate_step_euler(force, STEP_SIZE, vel[id], pos[id],
+                                           t);
+                } else if (integrator == integrator_t::RK4) {
+                  std::tie(wvel[id], wpos[id], std::ignore) =
+                      integrate_step_rk4(force, STEP_SIZE, vel[id], pos[id], t);
+                }
+              });
+        } break;
```
Comment (Ruyk) [inline demos/nbody/sim.hpp]: move this code out to separate functions

**Comment by rodburns** [comment]
I'll talk to Marcel about helping with making the changes. Thanks Ruyman.

File: demos/CMakeLists.txt [hunk 28]
```diff
@@ -0,0 +1,35 @@
+project (nbody)
+
+cmake_minimum_required(VERSION 3.2.2)
+set(CMAKE_VERBOSE_MAKEFILE ON)
+
+set(CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/../cmake/Modules")
+include(FindOpenCL)
+include(FindComputeCpp)
```
Comment (DuncanMcBain) [inline demos/CMakeLists.txt]: These should be find_package as of now.

File: demos/CMakeLists.txt [hunk 29]
```diff
@@ -0,0 +1,35 @@
...
+
+set(CMAKE_CXX_STANDARD 11)
+
+include("${CINDER_PATH}/proj/cmake/modules/cinderMakeApp.cmake")
+
+include_directories(SYSTEM "${COMPUTECPP_INCLUDE_DIRECTORY}")
+include_directories(SYSTEM "${OpenCL_INCLUDE_DIR}")
+include_directories(SYSTEM "${CINDER_PATH}/include")
+include_directories("${CMAKE_SOURCE_DIR}/include")
```
Comment (DuncanMcBain) [inline demos/CMakeLists.txt]: Start setting these per-target

File: demos/README.md [hunk 30]
```diff
@@ -0,0 +1,71 @@
...
+as well as libboost-system-dev and libboost-filesystem-dev on Ubuntu or their
+equivalents on other distributions. Then, clone two repositories into a folder
+of your choosing:
+
+```shell
+git clone --recursive https://github.com/cinder/Cinder
+git clone --recursive https://github.com/simongeilfus/Cinder-ImGui
+```
+
+In the Cinder folder (after `cd Cinder`), execute the following:
```
[the remark this answers is not in the document]
Comment (DuncanMcBain) [reply demos/README.md]: Check that this is still true, 1 year on

File: demos/README.md [hunk 31]
```diff
@@ -0,0 +1,71 @@
...
+```
+
+## Building
+After building Cinder, you can build the demos by running these commands in the
+`demos` folder of the SDK:
+
+```shell
+mkdir build
+cd build
+cmake -DCOMPUTECPP_PACKAGE_ROOT_DIR=/path/to/ccp/package/ \
```
Comment (DuncanMcBain) [inline demos/README.md]: Update these flags

File: demos/game_of_life/CMakeLists.txt [hunk 32]
```diff
@@ -0,0 +1,15 @@
...
+  "${CMAKE_CURRENT_SOURCE_DIR}/sim.cpp")
+
+ci_make_app(
+  APP_NAME "GameOfLife"
+  SOURCES "${SOURCES}"
+  LIBRARIES "${OpenCL_LIBRARIES}"
+  LIBRARIES "${COMPUTECPP_RUNTIME_LIBRARY}"
+  CINDER_PATH "${CINDER_PATH}")
+
+__build_spir("GameOfLife" "${CMAKE_CURRENT_SOURCE_DIR}/sim.cpp" "${CMAKE_CURRENT_BINARY_DIR}" 1)
```
Comment (DuncanMcBain) [inline demos/game_of_life/CMakeLists.txt]: We might be able to fix this with the new "add_sycl_to_target"

**Comment by Vtec234** [comment]
Hey, I looked at this again to try to adapt it to 0.9.0! It compiles correctly in [my branch here](https://github.com/Vtec234/computecpp-sdk/tree/demos-090), but a `sycl::exception` is thrown, claiming `One module without kernel function!`. This is confusing, because the `sim.cpp.sycl` file exists and has all the kernel info.

**Comment by DuncanMcBain** [comment]
Hey Wojciech! It's awesome you're updating this, we were planning to make some updates and squash it on the work you'd done here, thank you so much for adding to this MR! Don't feel in any way obliged to respond to the comments (because you're really really not), we wanted to leave them here as a log for others to pick up and work on.

**Comment by Vtec234** [comment]
Hey, that's no problem, I simply wanted to see if the demos still work with new SYCL :) And they do, so feel free to use the 0.9.0 version in [here](https://github.com/Vtec234/computecpp-sdk/tree/demos-090). I also added some functionality for Coulomb force solving and velocity visualisation to the n-body demo. I didn't push it here only because there are a lot of comments that people left which would become hidden and maybe a bit more annoying to review, but other than that the linked branch should be mergeable into the repo.

**Comment by keryell** [comment]
Actually I should have these flashy demo on my laptop too ! :-)

**Comment by DuncanMcBain** [comment]
Hi @Vtec234, what I'd like to do is take this work you've done, and the new functionality, and merge it up as soon as possible! It's clear this approach wasn't working and I think it'd be best now to get it in, then we can start actually using it properly, demoing it to people and so on. If you push your commits from the other branch, I'm happy to merge!

Thanks for your patience!

**Comment by Vtec234** [comment]
Hi, I changed the branch to the 0.9.1 version.

File: demos/include/integrator.hpp (hunk 7 above)
Comment (Vtec234) [reply demos/include/integrator.hpp]: I did write it, and I'm still not sure if it was a good idea or not.

**Pull request merged**